pkg/sentry/fsimpl/iouringfs/iouringfs.go

```diff
@@ -0,0 +1,323 @@
+	"gvisor.dev/gvisor/pkg/sentry/memmap"
+	"gvisor.dev/gvisor/pkg/sentry/pgalloc"
+	"gvisor.dev/gvisor/pkg/sentry/usage"
+	"gvisor.dev/gvisor/pkg/sentry/vfs"
+	"gvisor.dev/gvisor/pkg/syserror"
+)
+
+// filesystemType implements vfs.FilesystemType.
+//
+// +stateify savable
+type iouringFSType struct{}
+
+// GetFilesystem implements vfs.FilesystemType.GetFilesystem.
+func (iouringFSType) GetFilesystem(_ context.Context, vfsObj *vfs.VirtualFilesystem, _ *auth.Credentials, _ string, _ vfs.GetFilesystemOptions) (*vfs.Filesystem, *vfs.Dentry, error) {
+	panic("iouringfs.iouringFSType.GetFilesystem should never be called")
+}
+
+// Name implements vfs.FilesystemType.Name.
+//
+// Note that registering sockfs is unnecessary, except for the fact that it
+// will not show up under /proc/filesystems as a result. This is a very minor
+// discrepancy from Linux.
+func (iouringFSType) Name() string {
+	return "iouring"
+}
+
+// Release implements vfs.FilesystemType.Release.
+func (iouringFSType) Release(ctx context.Context) {}
+
+// +stateify savable
+type iouringFS struct {
+	kernfs.Filesystem
+
+	devMinor uint32
+}
+
+// NewFilesystem sets up and returns a new iouring filesystem.
+//
+// Note that there should only ever be one instance of iouring.Filesystem,
+// backing a global io_uring mount.
+func NewFilesystem(vfsObj *vfs.VirtualFilesystem) (*vfs.Filesystem, error) {
+	devMinor, err := vfsObj.GetAnonBlockDevMinor()
+	if err != nil {
+		return nil, err
+	}
+
+	fs := &iouringFS{
+		devMinor: devMinor,
+	}
+
+	fs.Filesystem.VFSFilesystem().Init(vfsObj, iouringFSType{}, fs)
+
+	return fs.Filesystem.VFSFilesystem(), nil
+}
+
+// Release implements vfs.FilesystemImpl.Release.
+func (fs *iouringFS) Release(ctx context.Context) {
+	fs.Filesystem.VFSFilesystem().VirtualFilesystem().PutAnonBlockDevMinor(fs.devMinor)
+	fs.Filesystem.Release(ctx)
+}
+
+// PrependPath implements vfs.FilesystemImpl.PrependPath.
+func (fs *iouringFS) PrependPath(ctx context.Context, vfsroot, vd vfs.VirtualDentry, b *fspath.Builder) error {
+	inode := vd.Dentry().Impl().(*kernfs.Dentry).Inode().(*inode)
+	b.PrependComponent(fmt.Sprintf("iouring:[%d]", inode.InodeAttrs.Ino()))
+	return vfs.PrependPathSyntheticError{}
+}
+
+// MountOptions implements vfs.FilesystemImpl.MountOptions.
+func (fs *iouringFS) MountOptions() string {
+	return ""
+}
+
+// inode implements kernfs.Inode.
+//
+// +stateify savable
+type inode struct {
+	kernfs.InodeAttrs
+	kernfs.InodeNoopRefCount
+	kernfs.InodeNotDirectory
+	kernfs.InodeNotSymlink
+}
+
+// Open implements kernfs.Inode.Open. iouringfs inode can not be opened
+func (i *inode) Open(ctx context.Context, rp *vfs.ResolvingPath, d *kernfs.Dentry, opts vfs.OpenOptions) (*vfs.FileDescription, error) {
+	return nil, syserror.EFAULT
+}
+
+// StatFS implements kernfs.Inode.StatFS.
+func (i *inode) StatFS(ctx context.Context, fs *vfs.Filesystem) (linux.Statfs, error) {
+	return vfs.GenericStatFS(linux.SOCKFS_MAGIC), syserror.EFAULT
```

Comment:
I needed to take a magic. And I chose SOCKFS because it seems to be the more appropriate one. Linux does not have a ``magic`` for `io_uring`. As ``Linux`` ``io_uring`` fd are backed by an anonymous file, maybe [``ANON_INODE_FS_MAGIC``](https://elixir.bootlin.com/linux/v5.10.50/source/include/uapi/linux/magic.h#L85) is more appropriate?

Reply:
No, I mean `SOCKFS_MAGIC` is appropriate, but you should document why in a comment nearby.

Reply:
Yep ok

Reply:
```
There is no IOURINGFS_MAGIC in `Linux` so we used `SOCKFS_MAGIC` here because `io_uring` is a message passing mechanism similar to sockets
```

Is this ok for you?